### ISSUE TITLE
rclone size & rclone about commands returns different results

### ISSUE THREAD

**Issue by honghainguyen777** [rclone/rclone#7924]
Hi Rclone community,

I am reaching out for assistance with an issue we're experiencing while migrating our OpenStack ObjectStore to a newer cluster. We encounter problems specifically with containers/buckets that contain a large number of files (> 100,000 files). Despite extensive research through Rclone GitHub Issues, forums, and other resources, we haven't found a viable solution.

#### Problem Description
We are trying to run the following command to migrate our data:
```
rclone copy source:my-bucket destination:my-bucket --transfers 64 --size-only --server-side-across-configs --multi-thread-write-buffer-size 3Mi --multi-thread-cutoff 1Gi --multi-thread-chunk-size 512Mi --progress -vv
```

We have experimented with various other options as well, but each time, only a portion of the bucket is copied, and no errors are reported.

Notably, the copied portion has the exact number of objects and size as reported by `rclone size source:my-bucket`:
```
root@root#rclone about source:my-bucket
Used:    24.596 GiB
Objects: 71.233k

root@root# rclone size source:my-bucket
Total objects: 47.978k (47978)
Total size: 12.043 GiB (12930961798 Byte)

root@root# rclone size destination:my-bucket
Total objects: 47.978k (47978)
Total size: 12.043 GiB (12930961798 Byte) 
```

The number of objects and size should match the output from `rclone about source:my-bucket`. This discrepancy leads us to suspect a potential bug or configuration issue.

#### Rclone Version
```
rclone v1.67.0
- os/version: ubuntu 20.04 (64 bit)
- os/kernel: 6.8.11-300.fc40.x86_64 (x86_64)
- os/type: linux
- os/arch: amd64
- go/version: go1.22.4
- go/linking: static
- go/tags: none
```

#### OS Details
```
NAME="Ubuntu"
VERSION="20.04.6 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.6 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
```

#### Cloud Storage System
OpenStack Swift - ObjectStore

#### Commands Tried
```
rclone copy/sync source:my-bucket destination:my-bucket --transfers 64 --size-only --server-side-across-configs --multi-thread-write-buffer-size 3Mi --multi-thread-cutoff 1Gi --multi-thread-chunk-size 512Mi --progress  -vv

rclone about source:my-bucket
rclone size source:my-bucket
rclone size destination:my-bucket
```

I would greatly appreciate any guidance or solutions to resolve this issue. If more information is needed, I am happy to provide it promptly.

Thank you very much for your support!

Best regards,
Hai

**Comment by honghainguyen777** [comment]
Hi, update here!

I discovered the issue with rclone size showing different results compared to rclone about. It turns out rclone isn't mapping the entire directory structure and is missing a significant portion of level-1 directories. `Rclone tree/lsd/lsjson` behaves the same.

Additionally, rclone struggles with very long paths, such as:

`f9/73/f97350101e1a9de922bdbb762a33695234102dc119ffafa25c15d0418ec82b23/@final/58/22/4ce0a556103a9a959af49edcbf6712936624f12bf27d8fce89fdfecb0ad6`

This results in many files being either "corrupted" (even though they are actually copied) or showing as "file not found in the destination" (even though the files do exist there).

To work around this, I used the `swift list` command to obtain a list of all objects, then extracted unique paths up to a specific pattern (e.g., up to the hash). After that, I ran `rclone copy` for each unique subfolder. This approach resolved the issue.

**Comment by ncw** [comment]
The fact that `rclone size` and `rclone copy` without `--fast-list` are showing the same objects is interesting too as `rclone size` will use a recursive listing by default, whereas `rclone copy` will recurse the directories manually, so a very different listing strategy.

Can you just check that

    rclone size source:my-bucket

and

    rclone size --disable ListR source:my-bucket

Really do show the same results.

> Additionally, rclone struggles with very long paths, such as:
> `f9/73/f97350101e1a9de922bdbb762a33695234102dc119ffafa25c15d0418ec82b23/@final/58/22/4ce0a556103a9a959af49edcbf6712936624f12bf27d8fce89fdfecb0ad6`

There are no length limits on anything in rclone so long paths shouldn't be a concern. What happened?

> This results in many files being either "corrupted" (even though they are actually copied) or showing as "file not found in the destination" (even though the files do exist there).

Odd!

> To work around this, I used the `swift list` command to obtain a list of all objects, then extracted unique paths up to a specific pattern (e.g., up to the hash). After that, I ran `rclone copy` for each unique subfolder. This approach resolved the issue.

Do you have directory markers? As in objects with content type `"application/directory"` possibly ending in a `/`?

> root@root#rclone about source:my-bucket
> Used:    24.596 GiB
> Objects: 71.233k
> 
> root@root# rclone size source:my-bucket
> Total objects: 47.978k (47978)
> Total size: 12.043 GiB (12930961798 Byte)

This discrepancy is very interesting. The data from `rclone about` is read from the container so is the storage systems view of the world. The data from `rclone size` is read by recursively listing the container.

So there are 24,964 objects missing and a total size missing of 13,478,792,105 bytes or 12.55 GiB

Does the bucket contain multipart uploads? And are the parts stored in the bucket itself rather than an auxillary (`_segments`) bucket?

> I discovered the issue with rclone size showing different results compared to rclone about. It turns out rclone isn't mapping the entire directory structure and is missing a significant portion of level-1 directories. `Rclone tree/lsd/lsjson` behaves the same.

If you do `rclone lsf source:my-bucket` do you see the correct top level directories? Or are there some missing there?

If you do `rclone lsf source:my-bucket -vv --dump bodies` does what the swift cluster sends match up with what rclone prints?

Thanks

Nick

**Comment by ncw** [comment]
In that response are you expecting 256 directories with 2 hex digits each? If so the server only sent part of them.

Rclone asked for up to 1000 directories and the server only gave back 50 or so and didn't indicate there were more.

From that it looks like the server is misbehaving somehow.

**Comment by honghainguyen777** [comment]
I am expecting to receive 136 two-digit directories. However, I received much lesser number of directories. Using `swift list` I can see all the objects nested in different folder levels without. I guess `swift list` performs recursively too.

Perhaps, as you said, the object store service seems to have some problems. I have to really go much deeper in the directory structure to be able to copy all objects to the new swift server.

Is there an option to tell Rclone to ask for more directories instead of up to 1000?

**Comment by ncw** [comment]
I managed to find out where this came up before - in https://github.com/ncw/swift/pull/167 - the author of that made a fix for rclone too in https://github.com/rclone/rclone/pull/5224 which somehow never got merged :-(

I've rebased that fix - you'll find it below.

Try adding `fetch_until_empty_page = true` to your config (or use the flag `--swift-fetch-until-empty-page`

[v1.68.0-beta.8050.8d80df2c8.fix-5224-swift-listings](https://beta.rclone.org/branch/fix-5224-swift-listings/v1.68.0-beta.8050.8d80df2c8.fix-5224-swift-listings/) on branch [fix-5224-swift-listings](https://github.com/rclone/rclone/tree/fix-5224-swift-listings) (uploaded in 15-30 mins)

**Comment by honghainguyen777** [comment]
Hi @ncw,

Thank you for releasing the new beta. I used the flag and got the correct size of the entire bucket now. I will try to sync and see :)

**Comment by ncw** [comment]
Thanks for testing.

I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.68

Let me know what happens and if there are more problems I'll re-open this issue.
